File: src/Wasp/Entity.hs
module Wasp.Entity
    ( Entity (..)
    , EntityField (..)
    , EntityFieldType (..)
    ) where

import qualified Data.Text as Text
import qualified Data.Aeson as Aeson
import Data.Aeson ((.=), object, ToJSON(..))


data Entity = Entity
    { entityName :: !String
    , entityFields :: ![EntityField]
    } deriving (Show, Eq)

data EntityField = EntityField
    { entityFieldName :: !String
    , entityFieldType :: !EntityFieldType
    } deriving (Show, Eq)

data EntityFieldType = EftString | EftBoolean deriving (Eq)

instance Show EntityFieldType where
    show EftString = "string"
    show EftBoolean = "boolean"

instance ToJSON Entity where
    toJSON entity = object
        [ "name" .= entityName entity
        , "fields" .= entityFields entity
        ]

instance ToJSON EntityField where
    toJSON entityField = object
        [ "name" .= entityFieldName entityField
        , "type" .= entityFieldType entityField
        ]

instance ToJSON EntityFieldType where
    toJSON = Aeson.String . Text.pack . show

File: waspc/data/Generator/templates/react-app/src/entities/_entity/_Entity.js
{{={= =}=}}
import uuidv4 from 'uuid/v4'

export default class {= entityClassName =} {
  _data = {}

  constructor (data = {}) {
    this._data = {
      id: data.id || uuidv4(),
      {=# entity.fields =}
      {= name =}: data.{= name =},
      {=/ entity.fields =}
    }
  }

  get id () {
    return this._data.id
  }

  {=# entity.fields =}
  get {= name =} () {
    return this._data.{= name =}
  }
  {=/ entity.fields =}

  toData () {
    return this._data
  }
}
